refactor apiGateway authorizers and permissions

dougmoscrop opened the pull request:
## What did you implement:

Splitting up: #2262
## How did you implement it:

This is the refactoring of apiGateway authorizers and permissions.
## How can we verify it:

`npm test` passes. I've also ran most of the integration tests.
## Todos:
- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

**_Is this ready for review?:**_ YES

dougmoscrop commented:
Hey @pmuens almost all done the split up. Resources and Methods will, obviously, be the bigger of them all.

pmuens commented:
Awesome @dougmoscrop! Thank you very much! 👍 💯 

Will take a deep dive into it ASAP!

RafalWilinski commented on lib/plugins/aws/deploy/compile/events/apiGateway/lib/authorizers.js:
```diff
@@ -5,103 +5,42 @@ const _ = require('lodash');
-
-          const normalizedAuthorizerName = authorizerName[0]
-              .toUpperCase() + authorizerName.substr(1);
+    _.forEach(this.validated.events, (event) => {
```
Why using underscore instead of `Array.forEach` in case it's an array, or `Object.keys` if it's object and you'd like to iterate over keys?

RafalWilinski commented on lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js:
```diff
@@ -4,80 +4,160 @@ const _ = require('lodash');
-    // validate that path and method exists for each http event in service
     _.forEach(this.serverless.service.functions, (functionObject, functionName) => {
-      functionObject.events.forEach(event => {
+      _.forEach(functionObject.events, (event) => {
```
Same here, I think `functionObject.events.forEach(event => ...)` is totally fine.

RafalWilinski replied:
Moreover, this whole loop can be refactored into `Array.map`. It's classic example of pushing to an array in forEach loop.

dougmoscrop commented:
It handles the case where it's undefined. If user has function with no
events, it blows up without checking.

Agree it's a left fold hidden in two loops. I will post a comment if the
other way iof doing it and you can decide.

On Fri, Oct 28, 2016, 7:24 AM Rafal Wilinski <notifications@github.com>
wrote:

> *@RafalWilinski* commented on this pull request.
> ------------------------------
>
> In lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js
> <https://github.com/serverless/serverless/pull/2510#pullrequestreview-6225182>
> :
>
> >      _.forEach(this.serverless.service.functions, (functionObject, functionName) => {
> -      functionObject.events.forEach(event => {
> +      _.forEach(functionObject.events, (event) => {
>
> Same here, I think functionObject.events.forEach(event => ...) is totally
> fine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/2510#pullrequestreview-6225182>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjzFDiFxkKt6HoCdxlAqlb53k1iR-4fks5q4dpHgaJpZM4KeAVg>
> .
>

dougmoscrop commented:
(however it is not array.map as an event produces 0 or 1 http events, and
each function produces 0 or N events. It's a reduce.)

On Fri, Oct 28, 2016, 8:23 AM Doug Moscrop <doug.moscrop@gmail.com> wrote:

> It handles the case where it's undefined. If user has function with no
> events, it blows up without checking.
>
> Agree it's a left fold hidden in two loops. I will post a comment if the
> other way iof doing it and you can decide.
>
> On Fri, Oct 28, 2016, 7:24 AM Rafal Wilinski <notifications@github.com>
> wrote:
>
> *@RafalWilinski* commented on this pull request.
> ------------------------------
>
> In lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js
> <https://github.com/serverless/serverless/pull/2510#pullrequestreview-6225182>
> :
>
> >      _.forEach(this.serverless.service.functions, (functionObject, functionName) => {
> -      functionObject.events.forEach(event => {
> +      _.forEach(functionObject.events, (event) => {
>
> Same here, I think functionObject.events.forEach(event => ...) is totally
> fine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/2510#pullrequestreview-6225182>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjzFDiFxkKt6HoCdxlAqlb53k1iR-4fks5q4dpHgaJpZM4KeAVg>
> .
>
>

dougmoscrop commented:
@RafalWilinski here's the comparison:

```javascript
const events = Object.keys(this.serverless.service.functions).reduce((memo, functionName) => {
  const functionObject = this.serverless.service.functions[functionName];
  if (Array.isArray(functionObject.events)) {
    functionObject.events.forEach((event) => {
      if (!_.has(event, 'http')) {
        return;
      }
      // parse event, same no matter what
      memo.push(/* event */);
    });
    return memo;
  });
```

could also do

```javascript
    const events = Object.keys(this.serverless.service.functions).reduce((memo, functionName) => {
      const functionObject = this.serverless.service.functions[functionName];
      if (!Array.isArray(functionObject.events)) {
        return memo;
      }

      return memo.concat(functionObject.events
        .filter((event) => _.has(event, 'http'))
        .map((event) => {
        /* ... */
        });
    });
```

vs two `_.forEach()` loops.

eahefnawy commented on lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js:
```diff
@@ -4,80 +4,160 @@ const _ = require('lodash');
+        this.serverless.service.getFunction(authorizer.name);
+        const normalizedName = authorizer.name[0].toUpperCase() + authorizer.name.substr(1);
+        arn = { 'Fn::GetAtt': [`${normalizedName}LambdaFunction`, 'Arn'] };
+        name = authorizer.name;
```
There seem to be some duplication here, could you maybe refactor it into one or two methods (ie `extractLambdaArn()` or something like that) and reuse that inside those if statements?

dougmoscrop commented:
Sure will do!

On Mon, Oct 31, 2016, 3:45 AM Eslam λ Hefnawy <notifications@github.com>
wrote:

> *@eahefnawy* requested changes on this pull request.
>
> @dougmoscrop <https://github.com/dougmoscrop> Thanks a lot! 😊 This looks
> much better than what we had before. Just added a tiny refactoring request
> to make this PR perfect.
> ------------------------------
>
> In lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js
> <https://github.com/serverless/serverless/pull/2510#pullrequestreview-6395187>
> :
>
> > +        arn = authorizer;
> +        const splitArn = arn.split(':');
> +        const splitLambdaName = splitArn[splitArn.length - 1].split('-');
> +        name = splitLambdaName[splitLambdaName.length - 1];
> +      }
> +    } else if (typeof authorizer === 'object') {
> +      if (authorizer.arn) {
> +        arn = authorizer.arn;
> +        const splitArn = arn.split(':');
> +        const splitLambdaName = splitArn[splitArn.length - 1].split('-');
> +        name = splitLambdaName[splitLambdaName.length - 1];
> +      } else if (authorizer.name) {
> +        this.serverless.service.getFunction(authorizer.name);
> +        const normalizedName = authorizer.name[0].toUpperCase() + authorizer.name.substr(1);
> +        arn = { 'Fn::GetAtt': [`${normalizedName}LambdaFunction`, 'Arn'] };
> +        name = authorizer.name;
>
> There seem to be some duplication here, could you maybe refactor it into
> one or two methods (ie extractLambdaArn() or something like that) and
> reuse that inside those if statements?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/2510#pullrequestreview-6395187>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjzFLnhDouN1SWDhvMuAIkx2o1ZN_e2ks5q5Zy1gaJpZM4KeAVg>
> .
>

dougmoscrop commented:
@eahefnawy should be good

eahefnawy commented:
Beautiful! just tested it out again. Works great!

Thanks a lot @dougmoscrop ... G2M 💥

dougmoscrop commented:
Thanks, woot! I will get on the next one which is the "real" stuff for
methods/resources.

On Wed, Nov 2, 2016 at 9:26 AM, Eslam λ Hefnawy <notifications@github.com>
wrote:

> Merged #2510 <https://github.com/serverless/serverless/pull/2510>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/2510#event-844880945>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAjzFIKNFhpu1o2AaLYMZMmH5IuG3nKNks5q6I-ZgaJpZM4KeAVg>
> .
>
